# Online Retail Store Customer Segmentation

 **Data Overview**

In [ ]:
#Import Library
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib

Using matplotlib backend: <object object at 0x7b5b249276a0>


In [ ]:
#Dispaly all the column from the dataframe
pd.pandas.set_option('display.max_columns',None)

In [ ]:
# Read the file
retaildata=pd.read_excel("/content/online_retail_II.xlsx")

In [ ]:
# Shape of the file
retaildata.shape

(525461, 8)

In [ ]:
# Highlights the Columns
retaildata.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [ ]:
# Information of the Columns
retaildata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [ ]:
# To view the Count of unique no of Description
retaildata["Description"].nunique()

4681

In [ ]:
# How many product are ther in each
retaildata["Description"].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    3549
REGENCY CAKESTAND 3 TIER              2212
STRAWBERRY CERAMIC TRINKET BOX        1843
PACK OF 72 RETRO SPOT CAKE CASES      1466
ASSORTED COLOUR BIRD ORNAMENT         1457
Name: Description, dtype: int64

In [ ]:
# Short the product which one is most purhased product along with the price
retaildata.groupby("Description").agg({"Quantity":"sum","Price":"sum"}).sort_values("Quantity", ascending=False).head(10)

Quantity     Price
Description                                           
WHITE HANGING HEART T-LIGHT HOLDER     57733  10897.48
WORLD WAR 2 GLIDERS ASSTD DESIGNS      54698    124.79
BROCADE RING PURSE                     47647     56.96
PACK OF 72 RETRO SPOT CAKE CASES       46106    972.25
ASSORTED COLOUR BIRD ORNAMENT          44925   2899.01
60 TEATIME FAIRY CAKE CASES            36326    909.30
PACK OF 60 PINK PAISLEY CAKE CASES     31822    822.99
JUMBO BAG RED RETROSPOT                30727   2897.17
SMALL POPCORN HOLDER                   29500    955.07
STRAWBERRY CERAMIC TRINKET BOX         26563   2661.45

The most purhased product is WHITE HANGING HEART T-LIGHT HOLDER.

In [ ]:
# Which product is most less purhased product
retaildata.groupby("Description").agg({"Quantity":"sum","Price":"sum"}).sort_values("Quantity", ascending=True).head(10)

Quantity  Price
Description                            
?                         -25373    0.0
given away                -20000    0.0
missing                   -15149    0.0
ebay sales                -13630    0.0
Zebra invcing error        -9000    0.0
Given away                 -8800    0.0
Ebay sales by the box.     -7100    0.0
wet damages                -4354    0.0
my error - connor          -4200    0.0
damages                    -3788    0.0

The most less purhased product is given away. so we take its as most retuned product.

In [ ]:
#  To view the null value
retaildata.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

The most of the null value come under form Customer id and Description. So i removed thes null value because the Description column has text and the customor id has unique.

In [ ]:
# By using Dropna to remove the null value
retaildata.dropna(inplace=True)

In [ ]:
# I can change the customer id from flot to int
retaildata["Customer ID"]=retaildata["Customer ID"].astype(int)
retaildata.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95        13085  United Kingdom  
1 2009-12-01 07:45:00   6.75        13085  United Kingdom  
2 2009-12-01 07:45:00   6.75        13085  United Kingdom  
3 2009-12-01 07:45:00   2.10        13085  United Kingdom  
4 2009-12-01 07:45:00   1.25        13085  United Kingdom

checking the most refunded product by controlling the refund invoices (invoices which starts with 'C')

again, we found that the most refunded product is paper craft little birdie

In [ ]:
#checking the most refunded product by controlling the refund invoices
retaildata_C= retaildata[retaildata["Invoice"].str.contains("C")==True]

In [ ]:
# In this retaildata_C dataframe we can groupby the Description and short the values based on Quantity
retaildata_C.groupby("Description").agg({"Quantity":"sum"}).sort_values("Quantity", ascending=True).head(10)

Quantity
Description                                  
ROTATING SILVER ANGELS T-LIGHT HLDR     -9363
SET/6 FRUIT SALAD PAPER CUPS            -7128
SET/6 FRUIT SALAD  PAPER PLATES         -7008
POP ART PEN CASE & PENS                 -5184
BLACK SILVER FLOWER T-LIGHT HOLDER      -5040
MULTICOLOUR SPRING FLOWER MUG           -4993
TEATIME PEN CASE & PENS                 -4632
WHITE BIRD GARDEN DESIGN MUG            -4320
S/4 BLUE ROUND DECOUPAGE BOXES          -3940
THE KING GIFT BAG                       -3744

I fine the most returned product is ROTATING SILVER ANGELS T-LIGHT HLDR.


In [ ]:
# To view how many invoice are there
retaildata["Invoice"].nunique()

23587

In [ ]:
# Find the toatal amount for how much money has been earned per invoice?
retaildata["TotalPrice"]=retaildata['Quantity']*retaildata['Price']

In [ ]:
#the invoice with the largest amount and its product
retaildata.groupby(["Invoice","Description"]).agg({"TotalPrice":"sum"}).sort_values("TotalPrice",ascending=False).head(10)

TotalPrice
Invoice Description                                    
530715  ROTATING SILVER ANGELS T-LIGHT HLDR    15818.40
502263  Manual                                 10953.50
524159  Manual                                 10468.80
496115  Manual                                  8985.60
511465  PINK PAPER PARASOL                      8925.00
501766  Manual                                  6958.17
501768  Manual                                  6958.17
525968  ROTATING SILVER ANGELS T-LIGHT HLDR     5179.20
524181  VINTAGE UNION JACK CUSHION COVER        4464.72
532358  ASSORTED COLOUR BIRD ORNAMENT           4176.00

The invoice with the largest amount and its product is 	"ROTATING SILVER ANGELS T-LIGHT HLDR"

In [ ]:
#The most expensive Products
retaildata.sort_values("Price", ascending=False).head(10)

Invoice StockCode Description  Quantity         InvoiceDate     Price  \
241824  C512770         M      Manual        -1 2010-06-17 16:52:00  25111.09   
135014  C502264         M      Manual        -1 2010-03-23 15:24:00  10953.50   
135013   502263         M      Manual         1 2010-03-23 15:22:00  10953.50   
135012  C502262         M      Manual        -1 2010-03-23 15:20:00  10953.50   
372833  C525398         M      Manual        -1 2010-10-05 11:47:00  10468.80   
358639   524159         M      Manual         1 2010-09-27 16:12:00  10468.80   
342135  C522793         M      Manual        -1 2010-09-16 14:53:00  10468.80   
74357   C496116         M      Manual        -1 2010-01-29 11:05:00   8985.60   
74356    496115         M      Manual         1 2010-01-29 11:04:00   8985.60   
374183  C525470         M      Manual        -1 2010-10-05 15:12:00   7044.79   

        Customer ID         Country  TotalPrice  
241824        17399  United Kingdom   -25111.09  
135014        12918  United Kingdom   -10953.50  
135013        12918  United Kingdom    10953.50  
135012        12918  United Kingdom   -10953.50  
372833        14063  United Kingdom   -10468.80  
358639        14063  United Kingdom    10468.80  
342135        14063  United Kingdom   -10468.80  
74357         17949  United Kingdom    -8985.60  
74356         17949  United Kingdom     8985.60  
374183        15413  United Kingdom    -7044.79

The most expensive Product and Price is "Manual" and "25111.09"

In [ ]:
# The product which we gained the most from,
retaildata.groupby("Description").agg({"TotalPrice":"sum"}).sort_values("TotalPrice",ascending=False).head(10)


TotalPrice
Description                                    
WHITE HANGING HEART T-LIGHT HOLDER    148876.66
REGENCY CAKESTAND 3 TIER              136866.30
ASSORTED COLOUR BIRD ORNAMENT          69854.96
JUMBO BAG RED RETROSPOT                51608.40
POSTAGE                                45520.86
ROTATING SILVER ANGELS T-LIGHT HLDR    39902.10
PAPER CHAIN KIT 50'S CHRISTMAS         36866.45
PARTY BUNTING                          34998.70
JUMBO BAG STRAWBERRY                   33893.70
VINTAGE UNION JACK BUNTING             33546.96

The product which we gained the most is **WHITE HANGING HEART T-LIGHT HOLDER**

In [ ]:
# Remove rows with invalid invoice that contains with C and assign it to a new data frame
retaildata['Invoice'] = retaildata['Invoice'].astype('str')
retaildata_2 = retaildata[~retaildata['Invoice'].str.contains('C')]
retaildata_2.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2009-12-01 07:45:00   6.95        13085  United Kingdom        83.4  
1 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
2 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
3 2009-12-01 07:45:00   2.10        13085  United Kingdom       100.8  
4 2009-12-01 07:45:00   1.25        13085  United Kingdom        30.0  
5 2009-12-01 07:45:00   1.65        13085  United Kingdom        39.6  
6 2009-12-01 07:45:00   1.25        13085  United Kingdom        30.0  
7 2009-12-01 07:45:00   5.95        13085  United Kingdom        59.5  
8 2009-12-01 07:46:00   2.55        13085  United Kingdom        30.6  
9 2009-12-01 07:46:00   3.75        13085  United Kingdom        45.0

In [ ]:
# How many products are there from each type
retaildata["Description"].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    3245
REGENCY CAKESTAND 3 TIER              1872
STRAWBERRY CERAMIC TRINKET BOX        1536
ASSORTED COLOUR BIRD ORNAMENT         1376
HOME BUILDING BLOCK WORD              1229
Name: Description, dtype: int64

In [ ]:
# How many orders came from which country we discard returns because they would be counted even if they were returned
retaildata_2["Country"].value_counts()

United Kingdom          370951
EIRE                      8507
Germany                   7661
France                    5470
Netherlands               2730
Spain                     1235
Switzerland               1170
Belgium                   1038
Portugal                   984
Sweden                     868
Channel Islands            821
Italy                      710
Australia                  630
Cyprus                     541
Austria                    524
Greece                     512
Denmark                    418
Norway                     365
Finland                    347
United Arab Emirates       315
Unspecified                277
USA                        230
Poland                     182
Malta                      170
Japan                      164
Lithuania                  154
Singapore                  117
Canada                      77
Thailand                    76
Israel                      74
Iceland                     71
RSA                         65
Brazil  

In [ ]:
# How much money we gained from the countries?
# we use retaildata again because the negative values have to drop from the total
retaildata.groupby("Country").agg({"TotalPrice":"sum"}).sort_values("TotalPrice",ascending=False).head(10)

TotalPrice
Country                    
United Kingdom  7038549.633
EIRE             328216.410
Netherlands      263863.410
Germany          196290.351
France           129773.830
Sweden            50859.510
Denmark           46972.950
Switzerland       43343.410
Spain             37084.900
Australia         30051.800

 The most money gained from the **United kingdom** and TotalPrice is **7038549.633 $**

# *Data Preparation*

In [ ]:
# Shape of the retaildata_2
retaildata_2.shape

(407695, 9)

In [ ]:
# Head columns of retaildata_2
retaildata_2.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2009-12-01 07:45:00   6.95        13085  United Kingdom        83.4  
1 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
2 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
3 2009-12-01 07:45:00   2.10        13085  United Kingdom       100.8  
4 2009-12-01 07:45:00   1.25        13085  United Kingdom        30.0

In [ ]:
retaildata_2.describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95, 0.99]).T

count          mean          std      min        1%        5%  \
Quantity     407695.0     13.586686    96.842229      1.0      1.00      1.00   
Price        407695.0      3.294188    34.756655      0.0      0.29      0.42   
Customer ID  407695.0  15368.504107  1679.795700  12346.0  12435.00  12731.00   
TotalPrice   407695.0     21.663261    77.147356      0.0      0.65      1.25   

                  10%       25%       50%       75%       90%      95%  \
Quantity         1.00      2.00      5.00     12.00     24.00     36.0   
Price            0.65      1.25      1.95      3.75      6.75      8.5   
Customer ID  13044.00  13997.00  15321.00  16812.00  17706.00  17913.0   
TotalPrice       2.10      4.95     11.90     19.50     35.70     67.5   

                  99%      max  
Quantity       144.00  19152.0  
Price           14.95  10953.5  
Customer ID  18196.00  18287.0  
TotalPrice     201.60  15818.4

In [ ]:
#we got the outlier values for our information, but we will not use it because we do not build a model
for feature in ["Quantity","Price","TotalPrice"]:

    Q1 = retaildata[feature].quantile(0.01)
    Q3 = retaildata[feature].quantile(0.99)
    IQR = Q3-Q1
    upper = Q3 + 1.5*IQR
    lower = Q1 - 1.5*IQR

    if retaildata[(retaildata[feature] > upper) | (retaildata[feature] < lower)].any(axis=None):
        print(feature,"yes")
        print(retaildata[(retaildata[feature] > upper) | (retaildata[feature] < lower)].shape[0])
    else:
        print(feature, "no")

Quantity yes
1063
Price yes
953
TotalPrice yes
1150


# Customer Segmentation with RFM Scores
RFM is abbreviation for Recency, Frequency and Monetary. It is a technique that helps determine marketing and sales strategies based on customers' buying habits. Recency: Time passed since the customer's last purchase. In other words, it is the "time since the last contact of the customer"

In [ ]:
retaildata_2.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2009-12-01 07:45:00   6.95        13085  United Kingdom        83.4  
1 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
2 2009-12-01 07:45:00   6.75        13085  United Kingdom        81.0  
3 2009-12-01 07:45:00   2.10        13085  United Kingdom       100.8  
4 2009-12-01 07:45:00   1.25        13085  United Kingdom        30.0

In [ ]:
retaildata_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 407695 entries, 0 to 525460
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      407695 non-null  object        
 1   StockCode    407695 non-null  object        
 2   Description  407695 non-null  object        
 3   Quantity     407695 non-null  int64         
 4   InvoiceDate  407695 non-null  datetime64[ns]
 5   Price        407695 non-null  float64       
 6   Customer ID  407695 non-null  int64         
 7   Country      407695 non-null  object        
 8   TotalPrice   407695 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 31.1+ MB


In [ ]:
retaildata_2["InvoiceDate"].min()

Timestamp('2009-12-01 07:45:00')

In [ ]:
retaildata_2["InvoiceDate"].max()

Timestamp('2010-12-09 20:01:00')

Now if we take date as the analyse date, then there will be a very serious difference.


For this reason, we have set the next day of the maximum day of the data set as the analyse date

In [ ]:
import datetime as dt
Last_date = dt.datetime(2010,12,10)
Last_date

datetime.datetime(2010, 12, 10, 0, 0)

In [ ]:
retaildata_2.groupby("Customer ID").agg({"InvoiceDate":"max"}).head(10)

InvoiceDate
Customer ID                    
12346       2010-06-28 13:53:00
12347       2010-12-07 14:57:00
12348       2010-09-27 14:59:00
12349       2010-10-28 08:23:00
12351       2010-11-29 15:23:00
12352       2010-11-29 10:07:00
12353       2010-10-27 12:44:00
12355       2010-05-21 11:59:00
12356       2010-11-24 12:24:00
12357       2010-11-16 14:29:00

what should we do now?

For each customer, we need to subtract the customers' last purchase date from this date.
And we will keep this as a temp data frame.

In [ ]:
(Last_date - retaildata_2.groupby("Customer ID").agg({"InvoiceDate":"max"})).head(10)

InvoiceDate
Customer ID                  
12346       164 days 10:07:00
12347         2 days 09:03:00
12348        73 days 09:01:00
12349        42 days 15:37:00
12351        10 days 08:37:00
12352        10 days 13:53:00
12353        43 days 11:16:00
12355       202 days 12:01:00
12356        15 days 11:36:00
12357        23 days 09:31:00

In [ ]:
temp_retaildata = (Last_date - retaildata_2.groupby("Customer ID").agg({"InvoiceDate":"max"}))

In [ ]:
(Last_date - retaildata_2.groupby(["Customer ID","Description"]).agg({"InvoiceDate":"max"})).head(60)
#I did this to check whether the Customer IDs are valid IDs with valid transactions

InvoiceDate
Customer ID Description                                          
12346       DOORMAT 3 SMILEY CATS               164 days 10:07:00
            DOORMAT AIRMAIL                     164 days 10:07:00
            DOORMAT BLACK FLOCK                 164 days 10:07:00
            DOORMAT CHRISTMAS VILLAGE           164 days 10:07:00
            DOORMAT FAIRY CAKE                  164 days 10:07:00
            DOORMAT FANCY FONT HOME SWEET HOME  164 days 10:07:00
            DOORMAT HOME SWEET HOME BLUE        164 days 10:07:00
            DOORMAT I LOVE LONDON               164 days 10:07:00
            DOORMAT MERRY CHRISTMAS RED         164 days 10:07:00
            DOORMAT NEW ENGLAND                 164 days 10:07:00
            DOORMAT PEACE ON EARTH BLUE         164 days 10:07:00
            DOORMAT RED SPOT                    164 days 10:07:00
            DOORMAT RESPECTABLE HOUSE           164 days 10:07:00
            DOORMAT SPOTTY HOME SWEET HOME      164 days 10:07:00
            DOORMAT UNION FLAG                  164 days 10:07:00
            DOORMAT UNION JACK GUNS AND ROSES   164 days 10:07:00
            DOORMAT WELCOME PUPPIES             164 days 10:07:00
            DOORMAT WELCOME SUNRISE             164 days 10:07:00
            DOORMAT WELCOME TO OUR HOME         164 days 10:07:00
            EDWARDIAN PARASOL BLACK             282 days 10:52:00
            EDWARDIAN PARASOL NATURAL           282 days 10:52:00
            EDWARDIAN PARASOL PINK              282 days 10:52:00
            EDWARDIAN PARASOL RED               282 days 10:52:00
            RED SPOTTY CHILDS UMBRELLA          282 days 10:52:00
            This is a test product.             321 days 10:30:00
12347       10 COLOUR SPACEBOY PEN               39 days 09:40:00
            3 HOOK HANGER MAGIC GARDEN           39 days 09:40:00
            3D DOG PICTURE PLAYING CARDS          2 days 09:03:00
            60 TEATIME FAIRY CAKE CASES          39 days 09:40:00
            72 SWEETHEART FAIRY CAKE CASES       39 days 09:40:00
            AIRLINE BAG VINTAGE JET SET BROWN     2 days 09:03:00
            ALARM CLOCK BAKELIKE CHOCOLATE        2 days 09:03:00
            ALARM CLOCK BAKELIKE GREEN            2 days 09:03:00
            ALARM CLOCK BAKELIKE ORANGE           2 days 09:03:00
            ALARM CLOCK BAKELIKE PINK             2 days 09:03:00
            ALARM CLOCK BAKELIKE RED              2 days 09:03:00
            BATHROOM METAL SIGN                   2 days 09:03:00
            BLACK CANDELABRA T-LIGHT HOLDER       2 days 09:03:00
            BLACK EAR MUFF HEADPHONES             2 days 09:03:00
            BLACK GRAND BAROQUE PHOTO FRAME       2 days 09:03:00
            BLUE 3 PIECE POLKADOT CUTLERY SET     2 days 09:03:00
            BLUE DRAWER KNOB ACRYLIC EDWARDIAN    2 days 09:03:00
            BLUE NEW BAROQUE CANDLESTICK CANDLE  39 days 09:40:00
            BOOM BOX SPEAKER BOYS                 2 days 09:03:00
            BOX OF 6 ASSORTED COLOUR TEASPOONS    2 days 09:03:00
            CAMOUFLAGE EAR MUFF HEADPHONES        2 days 09:03:00
            CAROUSEL DESIGN WASHBAG              39 days 09:40:00
            CLEAR DRAWER KNOB ACRYLIC EDWARDIAN   2 days 09:03:00
            COLOUR GLASS. STAR T-LIGHT HOLDER     2 days 09:03:00
            COSY SLIPPER SHOES LARGE GREEN       39 days 09:40:00
            COSY SLIPPER SHOES SMALL GREEN       39 days 09:40:00
            CUPID DESIGN SCENTED CANDLES         39 days 09:40:00
            EMERGENCY FIRST AID TIN               2 days 09:03:00
            FEATHER PEN,COAL BLACK               39 days 09:40:00
            FEATHER PEN,HOT PINK                 39 days 09:40:00
            FEATHER PEN,LIGHT PINK               39 days 09:40:00
            FOUR HOOK  WHITE LOVEBIRDS            2 days 09:03:00
            GREEN DRAWER KNOB ACRYLIC EDWARDIAN   2 days 09:03:00
            GREEN REGENCY TEACUP AND SAUCER      39 days 09:40:00
            HEART CALCULATOR

In [ ]:
#changing the name of InvoiceDate to Recency
temp_retaildata.rename(columns={"InvoiceDate": "Recency"}, inplace = True)

In [ ]:
temp_retaildata.head(10)

Recency
Customer ID                  
12346       164 days 10:07:00
12347         2 days 09:03:00
12348        73 days 09:01:00
12349        42 days 15:37:00
12351        10 days 08:37:00
12352        10 days 13:53:00
12353        43 days 11:16:00
12355       202 days 12:01:00
12356        15 days 11:36:00
12357        23 days 09:31:00

In [ ]:
# Applying Lambda Function
recency_retaildata = temp_retaildata["Recency"].apply(lambda x: x.days)

In [ ]:
recency_retaildata.head(10)

Customer ID
12346    164
12347      2
12348     73
12349     42
12351     10
12352     10
12353     43
12355    202
12356     15
12357     23
Name: Recency, dtype: int64

# Frequency

First, by grouping the data based on Customer ID and Invoice Number, we observe how many times each invoice is duplicated for each customer. Let's take a look at the first 20 observations.

In [ ]:
temp_retaildata = retaildata_2.groupby(["Customer ID","Invoice"]).agg({"Invoice":"count"})
temp_retaildata.head(20)

Invoice
Customer ID Invoice         
12346       491725         1
            491742         1
            491744         1
            492718         1
            492722         1
            493410         1
            493412         1
            494450         1
            495295         1
            499763         5
            513774        19
12347       529924        40
            537626        31
12348       524140        20
12349       506394        46
            508728         1
            529344        55
12351       536012        21
12352       532430         6
            535838        12

In [ ]:
#this shows how many different products each customer bought:
temp_retaildata.groupby("Customer ID").agg({"Invoice":"sum"}).head(10)

Invoice
Customer ID         
12346             33
12347             71
12348             20
12349            102
12351             21
12352             18
12353             20
12355             22
12356             84
12357            165

In [ ]:
#this shows how many times the customer did shopping:
temp_retaildata.groupby("Customer ID").agg({"Invoice":"count"}).head(10)

Invoice
Customer ID         
12346             11
12347              2
12348              1
12349              3
12351              1
12352              2
12353              1
12355              1
12356              3
12357              2

In [ ]:
#assigning this to a new dataframe called freq_df and changing the Invoice name to frequency
freq_retaildata = temp_retaildata.groupby("Customer ID").agg({"Invoice":"count"})
freq_retaildata.rename(columns={"Invoice": "Frequency"}, inplace = True)
freq_retaildata.head(10)

Frequency
Customer ID           
12346               11
12347                2
12348                1
12349                3
12351                1
12352                2
12353                1
12355                1
12356                3
12357                2

In [ ]:
# Sorting the values from the highest to the lowest. the customer with most orders:14911
freq_retaildata.sort_values("Frequency", ascending = False).head()

Frequency
Customer ID           
14911              205
17850              155
12748              144
15311              121
13089              109

In [ ]:
# Sorting the values from the highest to the lowest. the lowest customer's
freq_retaildata.sort_values("Frequency", ascending = True).head()

Frequency
Customer ID           
16178                1
14271                1
14269                1
14268                1
14266                1

# Monetary



Monetary is the total spend of the customer.

The sum of the TotalPrice values that we have calculated for each customer before is that customer's monetary value:

In [ ]:
monetary_retaildata = retaildata_2.groupby("Customer ID").agg({"TotalPrice":"sum"})
monetary_retaildata.head()

TotalPrice
Customer ID            
12346            372.86
12347           1323.32
12348            222.16
12349           2671.14
12351            300.93

In [ ]:
# Finding the customer which we gained the most money from: 18102 (the customer with most orders is 14911 so they are not the same)
monetary_retaildata.sort_values(by="TotalPrice", ascending=False).head(10)

TotalPrice
Customer ID            
18102         349164.35
14646         248396.50
14156         196566.74
14911         152147.57
13694         131443.19
17511          84541.17
15061          83284.38
16684          80489.21
16754          65500.07
17949          60117.60

In [ ]:
monetary_retaildata.rename(columns={"TotalPrice": "Monetary"}, inplace = True)
monetary_retaildata.head(10)

Monetary
Customer ID          
12346          372.86
12347         1323.32
12348          222.16
12349         2671.14
12351          300.93
12352          343.80
12353          317.76
12355          488.21
12356         3562.25
12357        12079.99

In [ ]:
print(recency_retaildata.shape, freq_retaildata.shape, monetary_retaildata.shape)

(4314,) (4314, 1) (4314, 1)


In [ ]:
#Lets bring all the values together under a new data frame called 'rfm':
rfm = pd.concat([recency_retaildata, freq_retaildata, monetary_retaildata],  axis=1)
rfm.head(10)

Recency  Frequency  Monetary
Customer ID                              
12346            164         11    372.86
12347              2          2   1323.32
12348             73          1    222.16
12349             42          3   2671.14
12351             10          1    300.93
12352             10          2    343.80
12353             43          1    317.76
12355            202          1    488.21
12356             15          3   3562.25
12357             23          2  12079.99

Normally the smallest of the recency scoring, which is 1, is the best recency score.

However, we will define this in reverse and put the value 5 as the best recency value so that it will be the same as the others,
so score 5 will be the most recent and the best recency score:

In [ ]:
rfm["RecencyScore"] = pd.qcut(rfm['Recency'], 5, labels = [5, 4, 3, 2, 1])

In [ ]:
rfm["Frequency"].describe([0.01,0.05,0.10,0.25,0.50,0.60,0.75,0.90,0.95, 0.99]).T

count    4314.000000
mean        4.454103
std         8.168658
min         1.000000
1%          1.000000
5%          1.000000
10%         1.000000
25%         1.000000
50%         2.000000
60%         3.000000
75%         5.000000
90%         9.000000
95%        13.000000
99%        31.000000
max       205.000000
Name: Frequency, dtype: float64

In [ ]:
cut_bins = [0,1,2,3,9,205]

rfm["FrequencyScore"] = pd.cut(rfm["Frequency"], bins = cut_bins, labels = [1, 2, 3, 4, 5])
rfm["FrequencyScore"].value_counts()

1    1421
4    1154
2     810
3     525
5     404
Name: FrequencyScore, dtype: int64

In [ ]:
rfm["MonetaryScore"] = pd.qcut(rfm['Monetary'], 5, labels = [1, 2, 3, 4, 5])
rfm.head(10)

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            164         11    372.86            2              5   
12347              2          2   1323.32            5              2   
12348             73          1    222.16            2              1   
12349             42          3   2671.14            3              3   
12351             10          1    300.93            5              1   
12352             10          2    343.80            5              2   
12353             43          1    317.76            3              1   
12355            202          1    488.21            1              1   
12356             15          3   3562.25            4              3   
12357             23          2  12079.99            4              2   

            MonetaryScore  
Customer ID                
12346                   2  
12347                   4  
12348                   1  
12349                   5  
12351                   2  
12352                   2  
12353                   2  
12355                   2  
12356                   5  
12357                   5

In [ ]:
(rfm['RecencyScore'].astype(str) +
 rfm['FrequencyScore'].astype(str) +
 rfm['MonetaryScore'].astype(str)).head(10)

Customer ID
12346    252
12347    524
12348    211
12349    335
12351    512
12352    522
12353    312
12355    112
12356    435
12357    425
dtype: object

In [ ]:
rfm["RFM_SCORE"] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str) + rfm['MonetaryScore'].astype(str)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            164         11    372.86            2              5   
12347              2          2   1323.32            5              2   
12348             73          1    222.16            2              1   
12349             42          3   2671.14            3              3   
12351             10          1    300.93            5              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   2       252  
12347                   4       524  
12348                   1       211  
12349                   5       335  
12351                   2       512

In [ ]:
#just to check whether 12346 is a test account or a valid customer account. it really is a valid account
retaildata[retaildata["Customer ID"]==12347].head()

Invoice StockCode                       Description  Quantity  \
423892  529924     22698    PINK REGENCY TEACUP AND SAUCER         6   
423893  529924     22699  ROSES REGENCY TEACUP AND SAUCER          6   
423894  529924     20985                  HEART CALCULATOR        12   
423895  529924     22418            10 COLOUR SPACEBOY PEN        24   
423896  529924     22422               TOOTHPASTE TUBE PEN        12   

               InvoiceDate  Price  Customer ID  Country  TotalPrice  
423892 2010-10-31 14:20:00   2.95        12347  Iceland        17.7  
423893 2010-10-31 14:20:00   2.95        12347  Iceland        17.7  
423894 2010-10-31 14:20:00   1.25        12347  Iceland        15.0  
423895 2010-10-31 14:20:00   0.85        12347  Iceland        20.4  
423896 2010-10-31 14:20:00   0.65        12347  Iceland         7.8

In [ ]:
rfm.describe().T

count         mean          std  min     25%     50%        75%  \
Recency    4314.0    90.269819    96.944304  0.0   17.00   52.00   135.0000   
Frequency  4314.0     4.454103     8.168658  1.0    1.00    2.00     5.0000   
Monetary   4314.0  2047.288659  8912.523243  0.0  307.95  705.55  1722.8025   

                 max  
Recency       373.00  
Frequency     205.00  
Monetary   349164.35

In [ ]:
#the champions:
rfm[rfm["RFM_SCORE"] == "555"].head(10)

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12431              8         13   4370.52            5              5   
12471              9         49  20139.74            5              5   
12472              4         13  11308.48            5              5   
12474             13         13   5048.66            5              5   
12483             13         11   2635.60            5              5   
12490             13         11   4405.71            5              5   
12523              9         12   2582.38            5              5   
12540             13         16   8854.25            5              5   
12583              8         13   7166.06            5              5   
12600              6         12   2444.88            5              5   

            MonetaryScore RFM_SCORE  
Customer ID                          
12431                   5       555  
12471                   5       555  
12472                   5       555  
12474                   5       555  
12483                   5       555  
12490                   5       555  
12523                   5       555  
12540                   5       555  
12583                   5       555  
12600                   5       555

In [ ]:
#the worst customers:
rfm[rfm["RFM_SCORE"] == "111"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12362            373          1    130.00            1              1   
12392            226          1    234.75            1              1   
12404            317          1     63.24            1              1   
12416            291          1    202.56            1              1   
12466            317          1     56.73            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12362                   1       111  
12392                   1       111  
12404                   1       111  
12416                   1       111  
12466                   1       111

In [ ]:
#To segment the customers using “Recency” and “Frequency” values:
#set up regular expression (regex) structure by using dictionaries to name Customer Segments according to
#Recency and Frequency Scores:

seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Lose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}

In [ ]:
#recency ve frequency points are obtained with the following code
#(as the monetary is a similar value to frequency, it is not considered here):
rfm['Segment'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)
rfm['Segment'].head()

Customer ID
12346    25
12347    52
12348    21
12349    33
12351    51
Name: Segment, dtype: object

In [ ]:
rfm['Segment'] = rfm['Segment'].replace(seg_map, regex=True)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            164         11    372.86            2              5   
12347              2          2   1323.32            5              2   
12348             73          1    222.16            2              1   
12349             42          3   2671.14            3              3   
12351             10          1    300.93            5              1   

            MonetaryScore RFM_SCORE              Segment  
Customer ID                                               
12346                   2       252           Can't Lose  
12347                   4       524  Potential Loyalists  
12348                   1       211          Hibernating  
12349                   5       335       Need Attention  
12351                   2       512        New Customers

In [ ]:
#Retrieve statistical values of recency, frequency and monetary values by segment classes:
rfm[["Segment", "Recency","Frequency","Monetary"]].groupby("Segment").agg(["mean","count"])

Recency        Frequency           Monetary      
                           mean count       mean count         mean count
Segment                                                                  
About to Sleep        52.726477   457   1.400438   457   523.735098   457
At Risk              137.770455   440   4.229545   440  1810.083161   440
Can't Lose           136.866667    15  16.866667    15  6166.613333    15
Champions              6.071656   628  13.085987   628  7195.632584   628
Hibernating          204.676000  1250   1.288800  1250   450.531211  1250
Loyal Customers       35.127820   665   7.273684   665  2952.225406   665
Need Attention        52.171875   128   3.000000   128  1340.991172   128
New Customers          8.101124    89   1.000000    89   337.436854    89
Potential Loyalists   17.243129   473   2.437632   473   842.126027   473
Promising             24.159763   169   1.000000   169   345.946805   169

In [ ]:
rfm[rfm["Segment"] == "Need Attention"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12349             42          3   2671.14            3              3   
12369             48          3   1791.15            3              3   
12371             44          3   2179.42            3              3   
12374             56          3   2246.29            3              3   
12389             37          3   1433.33            3              3   

            MonetaryScore RFM_SCORE         Segment  
Customer ID                                          
12349                   5       335  Need Attention  
12369                   4       334  Need Attention  
12371                   5       335  Need Attention  
12374                   5       335  Need Attention  
12389                   4       334  Need Attention

In [ ]:
need_attention_df = pd.DataFrame()
need_attention_df["Need_Attention_CustomerID"] = rfm[rfm["Segment"] == "Need Attention"].index
need_attention_df.head()

Need_Attention_CustomerID
0                      12349
1                      12369
2                      12371
3                      12374
4                      12389

In [ ]:
rfm[rfm["Segment"] == "At Risk"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12361             97          3    321.35            2              3   
12370            259          3    774.62            1              3   
12380            100          7   6951.49            2              4   
12391             99          5   2312.53            2              4   
12414             98          3    390.00            2              3   

            MonetaryScore RFM_SCORE  Segment  
Customer ID                                   
12361                   2       232  At Risk  
12370                   3       133  At Risk  
12380                   5       245  At Risk  
12391                   5       245  At Risk  
12414                   2       232  At Risk

In [ ]:
at_risk_df = pd.DataFrame()
at_risk_df["At_Risk_CustomerID"] = rfm[rfm["Segment"] == "At Risk"].index
at_risk_df.head()

At_Risk_CustomerID
0               12361
1               12370
2               12380
3               12391
4               12414

In [ ]:
rfm[rfm["Segment"] == "Can't Lose"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            164         11    372.86            2              5   
12482            211         29  23691.40            1              5   
13044            127         10   4293.43            2              5   
15013            122         10   1641.30            2              5   
15015            136         13   2365.02            2              5   

            MonetaryScore RFM_SCORE     Segment  
Customer ID                                      
12346                   2       252  Can't Lose  
12482                   5       155  Can't Lose  
13044                   5       255  Can't Lose  
15013                   4       254  Can't Lose  
15015                   5       255  Can't Lose

In [ ]:
cant_lose_df = pd.DataFrame()
cant_lose_df["Can't_Lose_CustomerID"] = rfm[rfm["Segment"] == "Can't Lose"].index
cant_lose_df.head()

Can't_Lose_CustomerID
0                  12346
1                  12482
2                  13044
3                  15013
4                  15015

In [ ]:
rfm[rfm["Segment"] == "Champions"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12360             14          5   1590.83            5              4   
12395              6          4   2048.64            5              4   
12415             10          7  19543.84            5              4   
12429              0          5   3735.51            5              4   
12431              8         13   4370.52            5              5   

            MonetaryScore RFM_SCORE    Segment  
Customer ID                                     
12360                   4       544  Champions  
12395                   4       544  Champions  
12415                   5       545  Champions  
12429                   5       545  Champions  
12431                   5       555  Champions

In [ ]:
Champions_df=pd.DataFrame()
Champions_df["Champions"]=rfm[rfm["Segment"]=="Champions"].index

In [ ]:
Champions_df.head()

Champions
0      12360
1      12395
2      12415
3      12429
4      12431

In [ ]:
rfm[rfm["Segment"] == "Hibernating"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12348             73          1    222.16            2              1   
12355            202          1    488.21            1              1   
12362            373          1    130.00            1              1   
12366            268          1    500.24            1              1   
12368            263          1    917.70            1              1   

            MonetaryScore RFM_SCORE      Segment  
Customer ID                                       
12348                   1       211  Hibernating  
12355                   2       112  Hibernating  
12362                   1       111  Hibernating  
12366                   2       112  Hibernating  
12368                   3       113  Hibernating

In [ ]:
Hibernating_df=pd.DataFrame()
Hibernating_df["Hibernating"]=rfm[rfm["Segment"]=="Hibernating"].index
Hibernating_df.head()

Hibernating
0        12348
1        12355
2        12362
3        12366
4        12368

In [ ]:
rfm[rfm["Segment"] == "New Customers"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12351             10          1    300.93            5              1   
12385             14          1   1938.40            5              1   
12386              1          1    258.90            5              1   
12427              6          1    303.50            5              1   
12441              1          1    173.55            5              1   

            MonetaryScore RFM_SCORE        Segment  
Customer ID                                         
12351                   2       512  New Customers  
12385                   4       514  New Customers  
12386                   2       512  New Customers  
12427                   2       512  New Customers  
12441                   1       511  New Customers

In [ ]:
New_Customers_df=pd.DataFrame()
New_Customers_df["New Customers"]=rfm[rfm["Segment"]=="New Customers"].index


New Customers
0          12351
1          12385
2          12386
3          12427
4          12441

In [ ]:
rfm[rfm["Segment"] == "Potential Loyalists"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12347              2          2   1323.32            5              2   
12352             10          2    343.80            5              2   
12356             15          3   3562.25            4              3   
12357             23          2  12079.99            4              2   
12358             10          3   2719.01            5              3   

            MonetaryScore RFM_SCORE              Segment  
Customer ID                                               
12347                   4       524  Potential Loyalists  
12352                   2       522  Potential Loyalists  
12356                   5       435  Potential Loyalists  
12357                   5       425  Potential Loyalists  
12358                   5       535  Potential Loyalists

In [ ]:
Potential_Loyalists_df=pd.DataFrame()
Potential_Loyalists_df["New Customers"]=rfm[rfm["Segment"]=="New Customers"].index

In [ ]:
Potential_Loyalists_df.head()

New Customers
0          12351
1          12385
2          12386
3          12427
4          12441

**CONCLUSION**
In my plan to segment customers based on behavior and develop customized marketing strategies is a smart move. By understanding and targeting distinct customer groups, the company can boost sales and customer satisfaction, allocate resources more effectively, and adapt its strategies as customer behaviors change, leading to long-term success.